org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/filesystem/JLSFsUtils.java

```diff
@@ -0,0 +1,173 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Microsoft Corporation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.filesystem;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import org.eclipse.core.internal.preferences.EclipsePreferences;
+import org.eclipse.core.resources.IContainer;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IProjectDescription;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IJavaProject;
+
+/**
+ * Utilities of the file system implementation.
+ */
+public class JLSFsUtils {
```

Comment:
I removed the fields' public modifiers and reserve for `generatesMetadataFilesAtProjectRoot`. Otherwise I'll get `java.lang.IllegalAccessError: class org.eclipse.jdt.ls.core.internal.filesystem.GradleProjectMetadataFileTest tried to access method 'boolean org.eclipse.jdt.ls.core.internal.filesystem.JLSFsUtils.generatesMetadataFilesAtProjectRoot()' (org.eclipse.jdt.ls.core.internal.filesystem.GradleProjectMetadataFileTest is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @61635670`

Reply:
Ah, yes. That needs to stay protected to permit the tests (that exist in the same named subpackage) to use it as if they existed in the same package.